helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java

```diff
@@ -111,41 +135,74 @@ public Response deleteMetadataStoreRealm(@PathParam("realm") String realm) {
+   * realm and that have the prefix.
+   *
+   * @param realm Query param in endpoint path: metadata store realm.
+   * @param prefix Query param in endpoint path: prefix substring of sharding key.
+   * @return Json representation for the sharding keys.
    */
   @GET
   @Path("/sharding-keys")
-  public Response getShardingKeys(@QueryParam("realm") String realm) {
-    Map<String, Object> responseMap;
-    Collection<String> shardingKeys;
+  public Response getShardingKeys(@QueryParam("realm") String realm,
+      @QueryParam("prefix") String prefix) {
     try {
-      // If realm is not set in query param, the endpoint is: "/sharding-keys"
-      // to get all sharding keys in a namespace.
-      if (realm == null) {
-        shardingKeys = _metadataStoreDirectory.getAllShardingKeys(_namespace);
-        // To avoid allocating unnecessary resource, limit the map's capacity only for
-        // SHARDING_KEYS.
-        responseMap = new HashMap<>(1);
+      if (realm == null && prefix == null) {
+        // For endpoint: "/sharding-keys" to get all sharding keys in a namespace.
+        return getAllShardingKeys();
+      } else if (prefix == null) {
+        // For endpoint: "/sharding-keys?realm={realm}"
+        return getAllShardingKeysInRealm(realm);
+      } else if (realm == null) {
+        // For endpoint: "/sharding-keys?prefix={prefix}"
+        return getAllShardingKeysUnderPath(prefix);
       } else {
-        // For endpoint: "/sharding-keys?realm={realmName}"
-        shardingKeys = _metadataStoreDirectory.getAllShardingKeysInRealm(_namespace, realm);
-        // To avoid allocating unnecessary resource, limit the map's capacity only for
-        // SHARDING_KEYS and SINGLE_METADATA_STORE_REALM.
-        responseMap = new HashMap<>(2);
-        responseMap.put(MetadataStoreRoutingConstants.SINGLE_METADATA_STORE_REALM, realm);
+        // For endpoint: "/sharding-keys?realm={realm}&prefix={prefix}"
+        return getRealmShardingKeysUnderPath(realm, prefix);
       }
     } catch (NoSuchElementException ex) {
       return notFound(ex.getMessage());
     }
+  }
 
-    responseMap.put(MetadataStoreRoutingConstants.SHARDING_KEYS, shardingKeys);
-
-    return JSONRepresentation(responseMap);
+  /**
+   * Gets all path-based sharding keys for a queried realm at endpoint:
+   * "GET /metadata-store-realms/{realm}/sharding-keys". This endpoint is equivalent to
+   * the endpoint: "GET /sharding-keys?realm={realm}".
+   *
+   * @param realm Queried metadata store realm to get sharding keys.
+   * @return All path-based sharding keys in the queried realm.
+   */
+  @GET
+  @Path("/metadata-store-realms/{realm}/sharding-keys")
+  public Response getRealmShardingKeys(@PathParam("realm") String realm) {
+    try {
+      return getAllShardingKeysInRealm(realm);
+    } catch (NoSuchElementException ex) {
+      return notFound(ex.getMessage());
```

Comment:
Given my understanding to the status code, `badRequest` is more suitable than `notFound` in this situation. What `notFound` entails is that this url is not matched to anything, however in this situation, the url is correctly connecting to the endpoint. Returning `notFound` might cause confusion. 
I understand that `badRequest` is not perfect in this case either, but it should be more suitable than `notFound`. If you have a better idea, I'm happy to discuss. (Ref: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)

Reply:
@NealSun96 Thanks for reviewing.
As you referred in the link, `400 Bad Request`: The request could not be understood by the server due to **malformed syntax**. If you request a POST/PUT with json payload which has syntax error, I believe `Bad Request` is more suitable.
`404 Not Found` means the endpoint is correct but resource is not found in server, which matches exactly our case here `NoSuchElementException`.  (refs as an examples: 
- https://stackoverflow.com/questions/26968479/404-not-found-or-bad-request;
- https://stackoverflow.com/questions/19671317/400-bad-request-http-error-code-meaning)

What do you think?

Reply:
`404` does not mean the endpoint is correct. If I enter an end point "/metadata-store-realms/test-realm/sharding-keys" and I receive a `404` in return, my first reaction would be to change "sharding-keys" to "sharding-key", then "shardingkeys". 
On a second thought, `404` with appropriate message does work better than `400`. `400` is not very fitting.

Reply:
In your example, the **endpoint** is `/metadata-store-realms/{realm}/sharding-keys` is already correct. But the resource `test-realm` does not exist, and 404 is a **correct** response code. If endpoint is incorrect, you would not get 404 but 400 (bad URI) instead.
So if you get 404, the first reaction is definitely checking the requested resource, not changing the endpoint.